File: Go/DesignMovieRentingSystem.go
//go:build ignore

package main

import (
	"fmt"
	"sort"
	"strconv"
	"strings"
)

// entry sorts by (price, shop) for available, and (price, shop, movie) for rented.
type entry struct{ price, shop, movie int }

func lessAvail(a, b entry) bool {
	if a.price != b.price {
		return a.price < b.price
	}
	return a.shop < b.shop
}

func lessRented(a, b entry) bool {
	if a.price != b.price {
		return a.price < b.price
	}
	if a.shop != b.shop {
		return a.shop < b.shop
	}
	return a.movie < b.movie
}

// insertSorted / removeSorted keep a slice ordered via binary search.
func insertSorted(s []entry, e entry, less func(a, b entry) bool) []entry {
	i := sort.Search(len(s), func(i int) bool { return !less(s[i], e) })
	s = append(s, entry{})
	copy(s[i+1:], s[i:])
	s[i] = e
	return s
}

func removeSorted(s []entry, e entry, less func(a, b entry) bool) []entry {
	i := sort.Search(len(s), func(i int) bool { return !less(s[i], e) })
	return append(s[:i], s[i+1:]...)
}

type MovieRentingSystem struct {
	available map[int][]entry
	rented    []entry
	price     map[[2]int]int
}

func Constructor(n int, entries [][]int) MovieRentingSystem {
	m := MovieRentingSystem{available: map[int][]entry{}, price: map[[2]int]int{}}
	for _, e := range entries {
		shop, movie, price := e[0], e[1], e[2]
		m.price[[2]int{shop, movie}] = price
		m.available[movie] = insertSorted(m.available[movie], entry{price, shop, movie}, lessAvail)
	}
	return m
}

func (m *MovieRentingSystem) Search(movie int) []int {
	s := m.available[movie]
	res := []int{}
	for i := 0; i < len(s) && i < 5; i++ {
		res = append(res, s[i].shop)
	}
	return res
}

func (m *MovieRentingSystem) Rent(shop, movie int) {
	p := m.price[[2]int{shop, movie}]
	e := entry{p, shop, movie}
	m.available[movie] = removeSorted(m.available[movie], e, lessAvail)
	m.rented = insertSorted(m.rented, e, lessRented)
}

func (m *MovieRentingSystem) Drop(shop, movie int) {
	p := m.price[[2]int{shop, movie}]
	e := entry{p, shop, movie}
	m.rented = removeSorted(m.rented, e, lessRented)
	m.available[movie] = insertSorted(m.available[movie], e, lessAvail)
}

func (m *MovieRentingSystem) Report() [][]int {
	res := [][]int{}
	for i := 0; i < len(m.rented) && i < 5; i++ {
		res = append(res, []int{m.rented[i].shop, m.rented[i].movie})
	}
	return res
}

func intList(xs []int) string {
	parts := make([]string, len(xs))
	for i, x := range xs {
		parts[i] = strconv.Itoa(x)
	}
	return "[" + strings.Join(parts, ", ") + "]"
}

func pairList(xs [][]int) string {
	parts := make([]string, len(xs))
	for i, p := range xs {
		parts[i] = intList(p)
	}
	return "[" + strings.Join(parts, ", ") + "]"
}

func main() {
	m := Constructor(3, [][]int{{0, 1, 5}, {0, 2, 6}, {0, 3, 7}, {1, 1, 4}, {1, 2, 7}, {2, 1, 5}})
	out := []string{"null"}
	out = append(out, intList(m.Search(1)))
	m.Rent(0, 1)
	out = append(out, "null")
	m.Rent(1, 2)
	out = append(out, "null")
	out = append(out, pairList(m.Report()))
	m.Drop(1, 2)
	out = append(out, "null")
	out = append(out, intList(m.Search(2)))
	fmt.Println("[" + strings.Join(out, ", ") + "]")
}

File: Java/DesignMovieRentingSystem.java
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.StringJoiner;
import java.util.TreeSet;

public class DesignMovieRentingSystem {
    static class MovieRentingSystem {
        // available[movie] = sorted set of {price, shop}; rented = sorted set of {price, shop, movie}.
        private final Map<Integer, TreeSet<int[]>> available = new HashMap<>();
        private final TreeSet<int[]> rented = new TreeSet<>(
                Comparator.<int[]>comparingInt(a -> a[0]).thenComparingInt(a -> a[1]).thenComparingInt(a -> a[2]));
        private final Map<Long, Integer> price = new HashMap<>();

        private static final Comparator<int[]> BY_PRICE_SHOP =
                Comparator.<int[]>comparingInt(a -> a[0]).thenComparingInt(a -> a[1]);

        MovieRentingSystem(int n, int[][] entries) {
            for (int[] e : entries) {
                int shop = e[0], movie = e[1], p = e[2];
                price.put(key(shop, movie), p);
                available.computeIfAbsent(movie, k -> new TreeSet<>(BY_PRICE_SHOP)).add(new int[] { p, shop });
            }
        }

        List<Integer> search(int movie) {
            List<Integer> res = new ArrayList<>();
            TreeSet<int[]> set = available.get(movie);
            if (set != null) {
                for (int[] e : set) {
                    if (res.size() == 5) break;
                    res.add(e[1]);
                }
            }
            return res;
        }

        void rent(int shop, int movie) {
            int p = price.get(key(shop, movie));
            available.get(movie).remove(new int[] { p, shop });
            rented.add(new int[] { p, shop, movie });
        }

        void drop(int shop, int movie) {
            int p = price.get(key(shop, movie));
            rented.remove(new int[] { p, shop, movie });
            available.get(movie).add(new int[] { p, shop });
        }

        List<List<Integer>> report() {
            List<List<Integer>> res = new ArrayList<>();
            for (int[] e : rented) {
                if (res.size() == 5) break;
                res.add(List.of(e[1], e[2]));
            }
            return res;
        }

        private static long key(int shop, int movie) {
            return (long) shop * 100000 + movie;
        }
    }

    public static void main(String[] args) {
        int[][] entries = { {0, 1, 5}, {0, 2, 6}, {0, 3, 7}, {1, 1, 4}, {1, 2, 7}, {2, 1, 5} };
        MovieRentingSystem m = new MovieRentingSystem(3, entries);
        StringJoiner sj = new StringJoiner(", ", "[", "]");
        sj.add("null");
        sj.add(m.search(1).toString());
        m.rent(0, 1);
        sj.add("null");
        m.rent(1, 2);
        sj.add("null");
        sj.add(m.report().toString());
        m.drop(1, 2);
        sj.add("null");
        sj.add(m.search(2).toString());
        System.out.println(sj);
    }
}
